stdlib/public/core/UnsafeBufferPointer.swift.gyb

```diff
@@ -648,27 +652,406 @@ extension Unsafe${Mutable}BufferPointer {
+  /// must be a trivial type.
+  ///
+  /// - Note: All buffer elements must already be initialized.
+  ///
+  /// - Parameters:
+  ///   - repeatedValue: The value used when updating this pointer's memory.
+  @inlinable // unsafe-performance
+  @_silgen_name("$sSr6assign9repeatingyx_tF")
+  public func update(repeating repeatedValue: Element) {
     guard let dstBase = _position else {
       return
     }
 
-    dstBase.assign(repeating: repeatedValue, count: count)
+    dstBase.update(repeating: repeatedValue, count: count)
   }
-  
+
+  @inlinable
+  @_alwaysEmitIntoClient
+  @available(*, deprecated, renamed: "update(repeating:)")
+  @_silgen_name("_swift_se0370_UnsafeMutableBufferPointer_assign_repeating")
+  public func assign(repeating repeatedValue: Element) {
+    update(repeating: repeatedValue)
+  }
+
+  /// Updates the buffer's initialized memory with the given elements.
+  ///
+  /// The buffer’s memory must be initialized or its `Element` type
+  /// must be a trivial type.
+  ///
+  /// - Parameter source: A sequence of elements to be used to update
+  ///   the buffer's contents.
+  /// - Returns: An iterator to any elements of `source` that didn't fit in the
+  ///   buffer, and the index one past the last updated element in the buffer.
+  @inlinable
+  @_alwaysEmitIntoClient
+  public func update<S: Sequence>(
+    from source: S
+  ) -> (unwritten: S.Iterator, index: Index) where S.Element == Element {
+    var iterator = source.makeIterator()
+    guard !self.isEmpty else { return (iterator, startIndex) }
+    _internalInvariant(_position != nil)
+    var index = startIndex
+    while index < endIndex {
+      guard let element = iterator.next() else { break }
+      _position._unsafelyUnwrappedUnchecked[index] = element
+      formIndex(after: &index)
+    }
+    return (iterator, index)
+  }
+
+  /// Updates the buffer's initialized memory with every element of the source.
+  ///
+  /// Prior to calling the `update(fromContentsOf:)` method on a buffer,
+  /// the first `source.count` elements of the buffer's memory must be
+  /// initialized, or the buffer's `Element` type must be a trivial type.
+  /// The buffer must reference enough initialized memory to accommodate
+  /// `source.count` elements.
+  ///
+  /// The returned index is one past the index of the last element updated.
+  /// If `source` contains no elements, the returned index is equal to the
+  /// buffer's `startIndex`. If `source` contains as many elements as the buffer
+  /// can hold, the returned index is equal to the buffer's `endIndex`.
+  ///
+  /// - Precondition: `self.count` >= `source.count`
+  ///
+  /// - Parameter source: A collection of elements to be used to update
+  ///   the buffer's contents.
+  /// - Returns: An index one past the index of the last element updated.
+  @inlinable
+  @_alwaysEmitIntoClient
+  public func update(
+    fromContentsOf source: some Collection<Element>
+  ) -> Index {
```

Comment:
This particular function is the most interesting case from the point of view of potential overlaps of the source and destination memory range. Since it works for any kind of source `Collection`, a pathological case can be devised where `source` is a lazy view over an `UnsafeBufferPointer`, which itself overlaps with with the end of the destination buffer. Obviously this would be a bug in functionality, but it wouldn't be unsafe (or undefined behaviour.)